### DISCUSSION TITLE
FAQ

### DISCUSSION THREAD

**Discussion by nothingismagick** [iotaledger/stronghold.rs#99] [Q&A]
Please use this discussion thread to post and discuss frequently asked questions about Stronghold.

**Comment by m-delvalle** [comment]
First things first:
- What is Stronghold? Why should I care?
- Is stronghold intended for end users? Or only for devs as a piece of a larger app?

**Reply by nothingismagick** [reply]
great first questions!

**Reply by nothingismagick** [reply]
Stronghold is an encrypted database that follows and encourages best practices from the perspective of security. If the software you use actually implements stronghold, it is a strong indicator that the engineering team behind that software made a wise decision.

Stronghold is a library, and as such is not itself intended for end users. It does not provide any type of GUI.

**Comment by Phyloiota** [comment]
Could it be also used maybe as a password manager in a normal windows computer? So a developer could buils a software solution out of those components that acts as a password or security manager on a normal PC system?

**Reply by nothingismagick** [reply]
Yes, but you don't want to store individual login secrets, because that violates the first principle of never revealing a secret. If you wanted to create a password manager, you would derive individual passwords from a primary secret using any number of methods. For example:

1. decrypt the primary secret `571ec49b416372c19b71f9949546eaa489816f20cda32c59fcaaa7fe28317a30`
2. use the primary secret to create a 64 byte signature for some string like `https://mail.google.com/mail/u/1`
3. use this signature as your password login string (perhaps truncated, or hashed, etc.)

This way you never store passwords again, just the addresses or "strings" that allow you to derive a password using a primary secret.

Obviously there are other ways you could do this, like:
1. decrypt the primary secret `correct-horse-battery-staple` 
2. append with your string `macos-login`
3. blake2b hash or argon2i that concatenated string to proper length

**Comment by Phyloiota** [comment]
SO - could we say Stronghold tries to provide something similar that Ledger Nano, etc does with hardware wallets? Absolutely disconnect the highest security encryption functions from the rest of the software environment? So actually a "software - hardware wallet"?

**Reply by nothingismagick** [reply]
While technically correct, the benefit of using a software approach is that we can put this software on a USB Armory that is unlocked with a Yubikey and you have a "homegrown" Nano. Further down the line, we expect to be able to place it in an FPGA so that the actual soft-cpu is auditable and highly performant.

**Comment by Phyloiota** [comment]
What makes it different and more secure than already existing solutions? What is the key point where you believe it succeeds over already existing stuff that is used and proven?

**Reply by nothingismagick** [reply]
Well, the main differentiators are: 
1. it is encrypted at rest, by design.
2. it keeps records encrypted in memory until they are needed by the runtime
3. it uses advanced memory protection techniques (where available by the OS)
4. it uses the actor model for process isolation
5. it has an internal p2p communication system for multiparty / multidevice security activities

The biggest selling point in my opinion, is that reading out a secret is discouraged (and not even available in our recommended client implementation), which renders some social engineering attacks moot: "hey, I am from the IF, give my your seed and I will recover your wallet for you". That won't work because the user cannot get to their seed, and can only use it by entering a password. Modern humans won't generally give out their password, even to spouses for example.

**Comment by Phyloiota** [comment]
Is it a "product" that is sold for a price, or a open source and free to use thing?

**Reply by nothingismagick** [reply]
Stronghold is an open source library that is free to use. Third parties are also free to develop their own "proprietary" uses of the technology.

**Comment by Phyloiota** [comment]
What is the downside of this high level of security? Does it slow devices down? Does it need a lot of resources and computational power the keep the system running? Anything that would influence user experience of the devide i.e. a smarthone? Battery usage or Ram usage?

**Reply by nothingismagick** [reply]
We are actively benchmarking all critical processes, and while there is a nominal degree of overhead involved in encryption & decryption, we feel that this is acceptable given the benefits. We are actively involved in e.g. enhancing the Riker open source Actor Model library, which is one of the primary performance bottlenecks. Obviously every application that runs on a device consumes system resources, but the alternative is trusting a centralized service to maintain and secure your secrets. In that case you are offloading your CO2 footprint to a service provider. We feel that the tradeoff here is worth it, and look forward to the Beta phase where we can provide some real benchmarks.

**Comment by Roy-export** [comment]
How is the data encrypted and is it auditable? Will you know if there is an unauthorised access?

**Reply by nothingismagick** [reply]
The [snapshot is encrypted](https://github.com/iotaledger/stronghold.rs/blob/e221dcb31519960e60982012da3c2ac154d989e1/engine/snapshot/src/logic.rs#L33) using `xchacha20poly1305`.

You cannot audit the snapshot unless it is decrypted. 

Each vault's record is also encrypted internally using [libsodium box encryption](https://github.com/iotaledger/stronghold.rs/blob/e221dcb31519960e60982012da3c2ac154d989e1/engine/vault/src/crypto_box.rs#L12), so it is impossible to merely hunt for all of the records in memory as plaintext.

Detecting unauthorized access, like for example attempting to decrypt the snapshot file, is quite hard because the entire file is encrypted and it would require adding another layer of monitoring which is outside of the purview of this library. I expect that would be a layer 2 implementation concern, but an interesting challenge.

**Reply by ABerghuis** [reply]
Having read some Q&A. Does the Stonghold database only consists of a couple of records with a id, and some hashed characters? So the smalest vault is like (0, Zmlyc3QgaGludAAAAAAAAAAAAAAAAAAA)?

**Comment by Phyloiota** [comment]
So you write: "for performing cryptographic operations in a secure computational zone that can be distributed across multiple devices"
Does this mean that if i have several devices (phone , laptop, tablet) all thse devices will be suplied with secure opeartions from a single "stronghold"? how could they share secrets and how is this sharing (assumably over wifi / bluetooth, etc) secured? Does it need to have an "user account" created for that?

**Reply by nothingismagick** [reply]
The primary use for this is to actually leverage the actor model and send the remote (trusted) stronghold a signing request, like e.g. a slip10 derivation, which upon completion will return a public key and a signature to the calling stronghold. If secrets need to be shared, in e.g. a multisig or split-key approach, they will be distributed using the noise protocol XX handshake ephemeral key for transit protection and then whatever additional encryption (asymmetric for example) the library implementor chooses.

The communication crate has been built using libp2p and websockets, but will be extended to bluetooth, NFC and air-gapped sneakernet.

**Comment by ggreeve** [comment]
I agree let’s start by answering the basics. Please describe some real use cases where this can be applied. From a consumer stand point. Thx

**Reply by nothingismagick** [reply]
It can be used as a DB for a crypto-currency wallet (and is being developed in tandem with Firefly)
It can be used for protecting secrets on a server (like what will be needed post coordicide).
It can be used for secret derivation and password sharing amongst teams.
It can be used to store Identity data and secure sharing of PII.

But the big thing to remember is that stronghold itself is a library - so use cases are only limited by the ingenuity of the engineering teams that implement it.

**Comment by Drahc-GitHub** [comment]
Is it necessary use Firefly to can use Stronghold?

**Reply by nothingismagick** [reply]
No, you do not have to use Firefly to use Stronghold.

**Comment by Drahc-GitHub** [comment]
It is a lot of work to make Stronghold possible. What's the "paid" that you " the IF" obtain if it can be used for free. What is the way that the IF is going to get some "profit s" for make such a hard work. What has incentivated you and the IF for making Stronghold? thx.

**Reply by nothingismagick** [reply]
Well, there is a lot to unpack here, but I will try. First of all, while many security engineers will advise you never to "roll your own crypto", what I think many people miss is that even if you can trust that something has been done properly and in accordance with a specification, you should always verify that it has been done correctly - and safely.

This brings me to my first point: open source software and security. By making Stronghold an open source library, it means that other engineers and security researchers are able to look at the source code and help us make it stronger. Secondly, the charter of the IOTA Foundation is such, that we have the explicit goals of not only making and maintaining open source software, but also helping to educate people. There is no feasible way our engineering department could consume the foundation's resources and create proprietary software. I am not a lawyer, but even if technically legal - I feel like it would be morally wrong.

Now, where are the profits? By building our own secure database, we OWN the issues and problems that may arise and are able to fix them. This means reduced time to resolution, which is a factor in development lifecycle costs. Next, since we have a high level of confidence in our software library, we can use it internally for securing other "products" of the foundation, like the wallet etc.

And finally, I think everyone involved in online transactions can agree, high quality open source security libraries make the world a safer place.

**Comment by GnrGnr** [comment]
Apart from any wallet developers and exchanges (i suppose), are there any other environments you see that would _particularly_ benefit from using Stronghold?

**Reply by nothingismagick** [reply]
We have built stronghold under the assumption that it will be particularly useful for identity-based projects, which will lend itself to a wide range of applications, such as SSO etc.

**Comment by GnrGnr** [comment]
Are there any plans to actively engage those who might benefit from using Stronghold, in order to get a broader feedback, contributions, collaboration (e.g. publications, active engagement, co-development, harbouring it e.g. under Eclipse stewardship, etc)? Is there anything the IOTA Community could help with to that regard?

**Reply by nothingismagick** [reply]
These are great questions that we will be actively addressing in the context of the Stronghold IOTA X-Teams. While Eclipse (or other foundation) stewardship is interesting, we are primarily focussed on stabilizing and auditing the software. You have to have something before you can acquire a steward.

**Reply by nothingismagick** [reply]
@GnrGnr - The X-Teams kickoff meeting for Stronghold will be Thursday, the 21st of January 2021 at 8pm CET.

**Comment by GnrGnr** [comment]
What are existing alternatives to Stronghold and in what aspects is Stronghold superior/where does it differ to those alternative solutions?

**Reply by nothingismagick** [reply]
During the planning phase we looked at a great range of prior art, and the list is pretty eye-watering. I need to prune that, collate them and make a comparison. Will come back to this question with a large table.

**Comment by TeeVeeEss** [comment]
Stronghold is created and maintained in the language Rust afaik. Are there plans to port it to other languages like Java, Python? How about using Stronghold on a website, will there be a JS-library?

**Reply by nothingismagick** [reply]
Yes, Stronghold is written and maintained in Stable Rust. There are plans to make foreign function interface (FFI) bindings to:
- C/C++
- nodejs (potentially deno)
- python
- go

We are unlikely to make a browserified stronghold, and similarly do not have faith in the linear memory model of WASM, but we are considering making a "stronghold lite" that is only the communication layer so the Browser can leverage a stronghold in a proper application or on another device.

**Reply by nothingismagick** [reply]
[update] It seems that there may be a community initiative to make the communications layer available for browsers (via classical JS or potentially WASM). It may arise out of the X-Teams, an EDF Grant, or both.

**Comment by Ti-m0** [comment]
What is so great about the actor model, especially in the context of stronghold? (in easy words ;))
What are well known applications of it?

**Reply by nothingismagick** [reply]
An actor model relies on message passing, instead of calling functions. This enhances security, but has the downside that in complicated flows you may have to manage state. Here is a little presentation I made about the actor model. 

[Actor Model Presentation](https://github.com/iotaledger/stronghold.rs/files/5878589/Actor.Model.Presentation_revised.pdf)

**Comment by Ti-m0** [comment]
From the X-Teams kick-off (Maik Piel):
Will Stronghold able to do all things like the gpg tools--> en/de-cryption, sign.....
Maybe it could maybe better integrated (think about email encryption with gpg)?

**Reply by nothingismagick** [reply]
Yes, the procedures can do anything with cryptography that you can dream up, assuming the underlying cryptographic primitives are available.

**Comment by Ti-m0** [comment]
From the X-Teams kick-off (Dr.Electron):
What's the plan for embedded HW? So use for example an ST microontroller. Would it be an abstraction layer and you would need to implement the hw specific things yourself or should it already be integrated fully?

**Reply by nothingismagick** [reply]
We are focussing right now on the USB Armory and integration with Yubikey. Then we are going to look into riscv architectures, fpga, and finally things like the esp20 and the st32 series.

Embedded systems are so incredibly complex and nuanced, I honestly don't see how such a small team can offer full integration - especially on systems that do not support rust natively and introducing rust-based C bindings becomes cumbersome. We will definitely putting more thought into this though.

**Comment by Ti-m0** [comment]
From the X-Teams kick-off (Xteams!!!!):
Could stronghold be used to build/secure something like the ledger nano harware wallet?

**Reply by nothingismagick** [reply]
Stronghold on a USB Armory with a Yubikey is very similar to a nano, both in performance and security. We have already proven that the Stronghold will build and bench on a USB Armory and can communicate with an Ubuntu host. Obviously the devil is in the details, but it is vey promising.

**Comment by allamedo** [comment]
Stronghold may be hosted on the cloud for multidevice use. Will firefly use this feature? Who will be hosting this instance if so?

**Reply by nothingismagick** [reply]
At the moment there are no immediate plans for Firefly to use this feature, however at some point it is expected that there will be an interface where you can granularly choose your stronghold. The first use of the communications system will be to use your mobile device's app to approve transactions for the desktop, and then we will go from there.

We expect that there will be opportunities for third-parties to offer this feature to the broader community, and that companies can use an internal service to enhance operational security.

**Comment by roukrak** [comment]
When will there be tutorials and how to's?
What I am trying to do is to encrypt a xrp wallet private key in the vault and sing a transaction with the your stronghold library.
Is this even possible? Is it possible to do this over commandline or even better in python (maybe stronghold is implemented in your client libraries)?

Thanks!

**Reply by felsweg-iota** [reply]
Hello @roukrak,

Stronghold is mainly a library and as such mainly focused on being integrated into other applications. [Wallet](https://github.com/iotaledger/wallet.rs) and [Identity](https://github.com/iotaledger/identity.rs) are two frameworks making use of Stronghold. However, we are currently working on updating the "products" inside the Stronghold repository, that also include a [CLI for Stronghold](https://github.com/iotaledger/stronghold.rs/issues/289). 

And to answer your answer other question: yes signing "something" ( eg. your transaction ) is possible. The main entry point are the cryptographic procedures mainly accessible through [Procedure calls in the interface](https://github.com/iotaledger/stronghold.rs/blob/65c1cc545025f4b9ecb7de90a58f8b47d7196fb7/client/src/interface.rs#L272), if you intent to write a small example yourself. 

I hope that helps. Let me know, if you have any more questions.

**Reply by felsweg-iota** [reply]
Some addendum: 

Tutorials, [documentation](https://wiki.iota.org/stronghold.rs/welcome) and examples are continuously updated. Please feel free to check here regularly :)

**Reply by roukrak** [reply]
Well I think I will have to learn some rust. I will tell you how it went and how I implemented it in my python code 👍🏻

Thank you for now!

**Reply by felsweg-iota** [reply]
Sure! Ask us any time.
